src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,166 @@
+#include "storage/table.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : AbstractOperator(in), _column_id{column_id}, _scan_type{scan_type}, _search_value(search_value) {}
+
+namespace {
+template <typename T, typename U = T>
+auto predicate_for_scantype(ScanType scan_type) {
+  switch (scan_type) {
+    case ScanType::OpEquals:
+      // Note: The unary "+" forces conversion to a function pointer - without it, the return type would be deduced to
+      //       be the type of the first lambda.
+      return +[](const T& lhs, const U& rhs) { return lhs == rhs; };
+    case ScanType::OpNotEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs != rhs; };
+    case ScanType::OpLessThan:
+      return +[](const T& lhs, const U& rhs) { return lhs < rhs; };
+    case ScanType::OpLessThanEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs <= rhs; };
+    case ScanType::OpGreaterThan:
+      return +[](const T& lhs, const U& rhs) { return lhs > rhs; };
+    case ScanType::OpGreaterThanEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs >= rhs; };
+    default:
+      Fail("Invalid Scan type.");
+  }
+}
+}  // namespace
+
+template <typename T>
+void TableScan::_scan_value_segment(const ChunkID chunk_id, const ValueSegment<T>& segment, PosList& pos_list) {
+  const auto& values = segment.values();
+  const auto segment_size = segment.size();
+  const auto predicate = predicate_for_scantype<T>(_scan_type);
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < segment_size; ++chunk_offset) {
+    if (!segment.is_null(chunk_offset) && !variant_is_null(_search_value) &&
+        predicate(values[chunk_offset], get<T>(_search_value))) {
+      pos_list.push_back({chunk_id, chunk_offset});
+    }
+  }
+}
+
+// Represents a ValueID that can be just between two integer numbers. This is used for comparing range scans of
+// ValueIDs in DictionarySegments where the cutoff for the scan does not have a ValueID assigned as it is not contained
+// in the dictionary.
+struct InBetweenValueID {
+  // clang-format off
+
+  friend bool operator==(const ValueID& other, const InBetweenValueID& self) {
+      return other == self.lower && self.is_contained();
+  }
+  friend bool operator!=(const ValueID& other, const InBetweenValueID& self) { return !(other == self); }
+  friend bool operator<(const ValueID& other, const InBetweenValueID& self) { return other < self.lower; }
+  friend bool operator<=(const ValueID& other, const InBetweenValueID& self) { return other < self.upper; }
+  friend bool operator>(const ValueID& other, const InBetweenValueID& self) { return other >= self.upper; }
+  friend bool operator>=(const ValueID& other, const InBetweenValueID& self) { return other >= self.lower; }
```

Comment:
The `friend` keyword here means, among other things, that we are defining an outside function, that is, defining the freestanding function `bool operator==(const ValueID&, const InBetweenValueID&)`. In particular, the `InBetweenValueID` (the struct we are currently in) is the _right_ hand side instead of what could be implemented by the member function `bool operator==(const ValueID&) const`. We did this to have all the functions tightly coupled here in the struct and to keep the interesting part (the differences between the functions) close together. Do you think another structure would be better?